ext/nnstreamer/tensor_filter/tensor_filter_llamacpp.cc

```diff
@@ -0,0 +1,350 @@
+  }
+
+  /** load dynamic backends */
+  ggml_backend_load_all ();
+  model_params = llama_model_default_params ();
+  model_params.n_gpu_layers = ngl;
+
+  model = llama_load_model_from_file ((char *) prop->model_files[0], model_params);
+  if (model == nullptr) {
+    throw std::invalid_argument ("Failed to load model");
+  }
+
+  ctx_params = llama_context_default_params ();
+
+  /** enable performance counters */
+  ctx_params.no_perf = false;
+
+  sampler_params = llama_sampler_chain_default_params ();
+  sampler_params.no_perf = false;
+  smpl = llama_sampler_chain_init (sampler_params);
+
+  llama_sampler_chain_add (smpl, llama_sampler_init_greedy ());
+}
+
+/**
+ * @brief Invoke llamacpp using input tensors
+ */
+void
+TensorFilterLlamaCpp::invoke (const GstTensorMemory *input, GstTensorMemory *output)
+{
+  UNUSED (input);
+  UNUSED (output);
+
+  throw std::runtime_error (
+      "llamacpp only supports invoke-dynamic mode. Set `invoke-dynamic=true`");
+}
+
+/**
+ * @brief Invoke llamacpp using input tensors
+ */
+void
+TensorFilterLlamaCpp::invoke_dynamic (GstTensorFilterProperties *prop,
+    const GstTensorMemory *input, GstTensorMemory *output)
+{
+  /** @todo: Get n_predict value from prop */
+  int n_prompt, n_pos, n_predict = 32, n_decode = 0;
+  llama_token new_token_id;
+  llama_batch batch;
+  std::string prompt, result = "";
+
+  UNUSED (prop);
+
+  prompt.assign ((char *) input[0].data, input[0].size);
+  prompt.erase (prompt.find_last_not_of (" \t\n\r\f\v") + 1);
+
+  n_prompt = -llama_tokenize (model, prompt.c_str (), prompt.size (), NULL, 0, true, true);
+
+  std::vector<llama_token> prompt_tokens (n_prompt);
+  if (llama_tokenize (model, prompt.c_str (), prompt.size (),
+          prompt_tokens.data (), prompt_tokens.size (), true, true)
+      < 0) {
+    throw std::invalid_argument ("Failed to tokenize the prompt");
+  }
+
+  ctx_params.n_ctx = n_prompt + n_predict - 1;
+  /** n_batch is the maximum number of tokens that can be processed in a single call to llama_decode */
+  ctx_params.n_batch = n_prompt;
+
+  ctx = llama_new_context_with_model (model, ctx_params);
+
+  if (ctx == nullptr) {
+    throw std::invalid_argument ("Failed to create llama context");
+  }
+
+  for (auto id : prompt_tokens) {
+    char buf[128];
+    int n = llama_token_to_piece (model, id, buf, sizeof (buf), 0, true);
+    if (n < 0) {
+      throw std::invalid_argument ("Failed to convert token to piece");
+    }
+    std::string s (buf, n);
+    result += s;
+  }
+
+  batch = llama_batch_get_one (prompt_tokens.data (), prompt_tokens.size ());
+
+  for (n_pos = 0; n_pos + batch.n_tokens < n_prompt + n_predict;) {
+    /** evaluate the current batch with the transformer model */
+    if (llama_decode (ctx, batch)) {
+      throw std::invalid_argument ("Failed to eval");
+    }
+
+    n_pos += batch.n_tokens;
+
+    /** sample the next token */
+    {
+      new_token_id = llama_sampler_sample (smpl, ctx, -1);
+
```

Comment:
Doesn't this function need to throw an exception on failure?

Reply:
Could I know what `failure` means?

Reply:
I'm wondering if a valid value is always passed.

Reply:
I couldn't find the part that validates the id value in several codes in the example 😥

Reply:
Okay :smile: